src/Microsoft.Health.Api/Features/HealthChecks/CachedHealthCheck.cs

```diff
@@ -35,52 +35,64 @@ internal sealed class CachedHealthCheck : IHealthCheck, IDisposable
             _healthCheckFactory = EnsureArg.IsNotNull(healthCheckFactory, nameof(healthCheckFactory));
             _options = EnsureArg.IsNotNull(options, nameof(options));
             _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+            _semaphore = new SemaphoreSlim(_options.MaxRefreshThreads, _options.MaxRefreshThreads);
         }
 
         public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
         {
-            DateTimeOffset currentTime = Clock.UtcNow;
-            if (IsCacheFresh(currentTime))
+            CacheSnapshot current = GetCacheSnapshot();
+            if (current.IsFresh && (_options.CacheFailure || current.Result.Status != HealthStatus.Unhealthy))
             {
-                return Task.FromResult(_cachedResult.Value);
+                return Task.FromResult(current.Result);
             }
 
-            return RefreshCache(context, currentTime, cancellationToken);
+            // If the cache is stale but not expired, then we make a best effort to refresh if the semaphore isn't
+            // currently occupied by another thread. Otherwise, we'll use the cached value.
+            TimeSpan maxWait = current.IsValid ? TimeSpan.Zero : Timeout.InfiniteTimeSpan;
+            return RefreshCache(context, current, maxWait, cancellationToken);
         }
 
-        private async Task<HealthCheckResult> RefreshCache(HealthCheckContext context, DateTimeOffset currentTime, CancellationToken cancellationToken)
+        public void Dispose()
         {
-            // If the cache is stale but not expired, then we make a best effort to refresh if the semaphore isn't
-            // currently occupied by another thread. Otherwise, we'll use the cached value.
-            TimeSpan maxWait = IsCacheValid(currentTime) ? TimeSpan.Zero : Timeout.InfiniteTimeSpan;
+            _semaphore.Dispose();
+            GC.SuppressFinalize(this);
+        }
 
+        private async Task<HealthCheckResult> RefreshCache(
+            HealthCheckContext context,
+            CacheSnapshot current,
+            TimeSpan maxSemaphoreWait,
+            CancellationToken cancellationToken)
+        {
             try
             {
                 // Note: Only the timespan controls whether or not the method returns true/false.
                 //       If the token is canceled, the method throws instead an OperationCanceledException.
-                if (!await _semaphore.WaitAsync(maxWait, cancellationToken).ConfigureAwait(false))
+                if (!await _semaphore.WaitAsync(maxSemaphoreWait, cancellationToken).ConfigureAwait(false))
```

Comment:
There are so many try catch here, it is hard to follow. Can this be composed to something like

GetHealthCheckFromOrigin() // Get actual sql health check.

RefreshCache()
{
    semaphore.Wait();
    try
    {
        GetHealthCheckFromOrigin();
        UpdateCache();
    }
    finally
    {
        semaphore.Release();
    }
}

Reply:
This wait is a bit tricky. The code path for exiting via timeout is when the cache hasn't expired. Threads will attempt to enter the semaphore and give up quickly if someone else is already refreshing. Exiting via exception means that the caller has abandoned waiting on the health check entirely